### PR TITLE
Adding database

### PR DESCRIPTION
**Pull request by caesarx26**
# Pull Request Template

## Description

I have initialized making the database and making tests for the database help functions. There are no tests for the db or pages that use the db helper functions.

This is the migration used to make the tables:
```  TypeScript
// Function to initialize the database schema
export async function migrateDbIfNeeded() {
  const db = SQLite.openDatabaseAsync('bookCollection.db');

  await (await db).execAsync(`
        PRAGMA journal_mode = WAL;

        -- Create categories table with name as the primary key 
        CREATE TABLE IF NOT EXISTS categories (
          name TEXT PRIMARY KEY NOT NULL,
          isPinned INTEGER NOT NULL,
          position INTEGER NOT NULL
        );
        
        -- Create userBooks table to store user's books: id will be uuid generated for book when it is added to db 
        CREATE TABLE IF NOT EXISTS userBooks (
          id TEXT PRIMARY KEY NOT NULL,
          title TEXT NOT NULL,
          author TEXT NOT NULL,
          excerpt TEXT,
          summary TEXT,
          image TEXT,
          rating INTEGER,
          genres TEXT,
          category TEXT,
          isFavorite INTEGER,
          publisher TEXT,               
          publishedDate TEXT,           
          pageCount INTEGER,
          isCustomBook INTEGER,
          isbn TEXT,
          notes TEXT          
        );

        -- Create recommendedBooks table to store recommended books: id will be uuid generated for book when it is added to db
        CREATE TABLE IF NOT EXISTS recommendedBooks (
          id TEXT PRIMARY KEY NOT NULL,
          title TEXT NOT NULL,
          author TEXT NOT NULL,
          excerpt TEXT,
          summary TEXT,
          image TEXT,
          rating INTEGER,
          genres TEXT,
          addToLibrary INTEGER,
          publisher TEXT,               
          publishedDate TEXT,           
          pageCount INTEGER,
          isbn TEXT,
          notes TEXT             
        );
      `);

  console.log("Database initialized or migrated with new schema");
}
```

- There are helper functions are in database/databaseOperations.ts
- when books are made a uuid is generated as the id and added to the db with that id. This allows for duplicates and for books to be the same but with different categories. 
- These helper functions make querys to get get books or categories and to add/update books or categories
- I also deleted some components that were not being used and organized them better 
- I also moved icon components out of the asset folder into components folder (you can also import SVG files directly now as components this was done in a previous PR)

## Type of Change

- [ x] Bugfix
- [x ] New Feature
- [ ] Documentation Update
- [ ] Other (please describe):

## Checklist

- [ x] I have read the contributing guidelines.
- [ x] I have followed the code style of this project.
- [ x] I have added tests to cover my changes.
- [ x] I have updated the documentation (if applicable).
- [ ] My changes do not generate new warnings.

## Related Issue

#44 
#36 
#19 

## Additional Notes
Also implemented adding a custom book button still needs some modifying and also set a fixed height for the footer to avoid any visual glitches

### CODE REVIEW

File: Tabby/app/library/[category]/index.tsx [hunk 1]
```diff
@@ -1,12 +1,12 @@
 import React, { useState } from 'react';
-import { FlatList, Pressable } from 'react-native';
+import { FlatList, Pressable, Modal, TextInput, Button, View, Text, Image } from 'react-native';
```
Comment (LilAttacker) [inline Tabby/app/library/[category]/index.tsx:2]: Github Actions raises some issues here.

File: Tabby/app/library/[category]/index.tsx [hunk 2]
```diff
@@ -15,9 +15,11 @@ type Book = {
     isFavorite: boolean;
 };
 
+const defaultImage = require('@/assets/book/default-book-cover.jpg');
```
Comment (LilAttacker) [inline Tabby/app/library/[category]/index.tsx:18]: Github Actions also raises a warning here.

File: Tabby/app/testingdb.tsx [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import { Text } from 'react-native';
+import { SafeAreaView } from 'react-native-safe-area-context';
+import { Link } from 'expo-router';
```
Comment (LilAttacker) [inline Tabby/app/testingdb.tsx:4]: There's a few Github warnings here.

**Comment by LilAttacker** [comment]
Overall, I actually quite like this so far and think this would be an excellent addition once this is done. I just raised some concerns that Gthub Actions warned about.

File: Tabby/database/testMigration.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      title TEXT NOT NULL,
+      author TEXT NOT NULL,
+      excerpt TEXT,
+      summary TEXT,
+      image TEXT,
+      rating INTEGER,
+      genres TEXT,
+      isFavorite INTEGER,
+      addToLibrary INTEGER
+    );
```
Comment (LilAttacker) [inline Tabby/database/testMigration.ts:44]: If these are to be the tables that stores all frontend user information, labelling these tables with their purpose I think is a good idea.

**Comment by David1000734** [comment]
So I suppose the frontend will create a SQL file in the user's phone to store information for the books. Do we know if every phone will have the capability of running SQL queries?

**Comment by LilAttacker** [comment]
Just wanted to say that couple of warnings were raised by GitHub Actions and also I notice a distinct lack in documentation so far.

**Comment by LilAttacker** [comment]
Could you provide an updated description of what this PR does now? A lot of changes have been added and done, so knowing the latest changes would be nice.

**Comment by caesarx26** [comment]
> So I suppose the frontend will create a SQL file in the user's phone to store information for the books. Do we know if every phone will have the capability of running SQL queries?

yes

**Review by David1000734 — APPROVED** [review]
Looks good. I think it's an interesting choice to use TEXT for some values like isPinned that could be BOOLEAN instead but its fine

**Comment by caesarx26** [comment]
For sqlite there is no Boolean. With text you can also store long strings. > Looks good. I think it's an interesting choice to use TEXT for some values like isPinned that could be BOOLEAN instead but its fine

**Comment by caesarx26** [comment]
> Looks good. I think it's an interesting choice to use TEXT for some values like isPinned that could be BOOLEAN instead but its fine

For sqlite there is no Boolean. With text you can also store long strings.

**Review by LilAttacker — APPROVED** [review]
I believe the initial books and categories should be removed, but its fine to keep for the last minute testing this requires.

My only real issue is the lack of test coverage, but this feature is too important to delay any longer.

"We can test in production. That never goes wrong."

**Comment by MrSnoodle01** [comment]
Everything here looks good, I would like to see more tests but we are kind of running out of time at the moment, so it is more important to get the feature integrated with the rest of the app. I also like all of the helper functions, having that many should make adding this feature into the app a seamless process.

**Review by MrSnoodle01 — APPROVED** [review]

**Pull request merged**